### PR TITLE
added support for custom namespaced class names in associations

### PR DESCRIPTION
**Pull request by amicojeko**
now you can add associations with custom namespaces

`has_many :countries, class_name: '::Country'`

### CODE REVIEW

**Comment by lleirborras** [comment]
Nice! @vjt we should merge it immediately

File: lib/hawk/model/association.rb [hunk 1]
```diff
@@ -73,17 +73,16 @@ module ClassMethods
         def inherited(subclass)
           super
 
-          parent = self
           subclass.instance_eval do
             # Inherit associations
             @_associations ||= {}
 
-            parent.associations.each do |name, (type, options)|
```
Comment (vjt) [inline lib/hawk/model/association.rb]: As discussed verbally, here the `parent` class is saved in a closure in order to copy the associations from the parent in the inherited class. If you remove it, then ..... ------>

File: lib/hawk/model/association.rb [hunk 2]
```diff
@@ -73,17 +73,16 @@ module ClassMethods
...
+            self.associations.each do |name, (type, options)|
```
Comment (vjt) [inline lib/hawk/model/association.rb]: -----> you would then try copying the associations from yourself, and end up with an empty associations list :smile:

File: lib/hawk/model/association.rb [hunk 3]
```diff
@@ -230,7 +231,7 @@ def #{entities}
                 return @_#{entities} if instance_variable_defined?('@_#{entities}')
                 params = clean_inherited_params(self.params, #{conditions})
 
-                @_#{entities} = #{parent}::#{klass}.where(params)
+                @_#{entities} = #{klass_name}.where(params)
```
Comment (vjt) [inline lib/hawk/model/association.rb]: This is getting really :spaghetti: .. these `eval` blocks need refactoring in separate reusable micro-methods.

File: lib/hawk/model/association.rb [hunk 4]
```diff
@@ -73,17 +73,16 @@ module ClassMethods
...
           super
 
-          parent = self
           subclass.instance_eval do
             # Inherit associations
             @_associations ||= {}
 
-            parent.associations.each do |name, (type, options)|
+            self.associations.each do |name, (type, options)|
               _define_association(name, type, options)
```
Comment (vjt) [inline lib/hawk/model/association.rb]: Here, as discussed, we should take care of mangling the `options[:class_name]` if we detect being in a namespace different than the one we are inheriting from.

THANKS

**Comment by vjt** [comment]
Sorry but my **OCD** prevents me merging, despite beautifully constructed, blatantly wrong changes :smile: :smile: :smile:

File: lib/hawk/model/association.rb [hunk 5]
```diff
@@ -79,6 +79,17 @@ def inherited(subclass)
...
+
+              namespaced = Object.const_get(self_constant) rescue nil
+
+              options[:class_name] = if namespaced && (parent.to_s.deconstantize != self.to_s.deconstantize)
+                self_constant
+              else
+                parent_constant
+              end
+
               _define_association(name, type, options)
```
Comment (vjt) [inline lib/hawk/model/association.rb]: Here we could use some code comments, such as *The idea is to look up this class' name in the same namespace*

**Comment by vjt** [comment]
Well done guys. This is a good improvement. Thanks :beers:  :smile_cat:

**Pull request merged**